src/asyncUtils.ts

```diff
@@ -33,26 +40,32 @@ function asyncUtils(addResolver) {
           resolve()
         })
       })
-      await act(() => nextUpdatePromise)
+      await act(() => nextUpdatePromise as Promise<void>)
     }
     await nextUpdatePromise
   }
 
-  const waitFor = async (callback, { interval, timeout, suppressErrors = true } = {}) => {
-    // eslint-disable-next-line consistent-return
+  // TODO: Discuss with Kent and Maintainers about behavior of returning nothing currently there are tests handling this behavior that may be an anti-pattern.
+  // ? Should waitFor() always expect something returned
+  const waitFor = async <T>(
+    callback: () => T | Promise<T>,
+    { interval, timeout, suppressErrors = true }: WaitOptions = {}
+  ) => {
     const checkResult = () => {
       try {
         const callbackResult = callback()
         return callbackResult || callbackResult === undefined
-      } catch (e) {
+      } catch (error: unknown) {
         if (!suppressErrors) {
-          throw e
+          throw error as Error
         }
+        return undefined
       }
     }
```

Comment:
@kentcdodds @mpeyper This is the current Types for `waitFor()` the type is currently expecting "something" to be returned. We could make it more explicit to `never` expect `void` `null` etc...

Reply:
There are two use cases for `waitFor` in this library (I'm not sure how it behaves in RTL)

1. `waitFor` this thing to not throw an exception:
   ```js
   waitFor(() => {
     expect(result.current).toBe(expectedValue)
   })
   ```
   In this case, the only value that would break the waiting is if the callback returns `undefined` (no return).
2. `waitFor` this thing to return `truthy`
   ```js
   waitFor(() => result.current.state === expectedValue)
   ```
   In this case, the wait will only break when the `true` (or any `thruthy` value) is returned.

Does that help?

Reply:
Definitely helps! Thanks.